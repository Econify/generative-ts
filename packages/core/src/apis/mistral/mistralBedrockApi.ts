import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/lib/Either.js";

import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { EjsTemplate } from "../../utils/ejsTemplate";

import type { FewShotRequestOptions } from "../shared";

import { Llama2ChatMlTemplateSource } from "../meta/llama2ChatApi";

const templateSource = `{
  "prompt": "${Llama2ChatMlTemplateSource}"
  <% if (typeof max_tokens !== 'undefined') { %>, "max_tokens": <%= max_tokens %><% } %>
  <% if (typeof stop !== 'undefined') { %>, "stop": [<%= stop.join(', ') %>]<% } %>
  <% if (typeof temperature !== 'undefined') { %>, "temperature": <%= temperature %><% } %>
  <% if (typeof top_p !== 'undefined') { %>, "top_p": <%= top_p %><% } %>
  <% if (typeof top_k !== 'undefined') { %>, "top_k": <%= top_k %><% } %>  
}`;

/**
 * @category Requests
 * @category Mistral Bedrock
 */
export interface MistralBedrockOptions
  extends FewShotRequestOptions,
    ModelRequestOptions {
  max_tokens?: number;
  stop?: string[];
  temperature?: number;
  top_p?: number;
  top_k?: number;
}

/**
 * @category Templates
 * @category Mistral Bedrock
 */
export const MistralBedrockTemplate = new EjsTemplate<MistralBedrockOptions>(
  templateSource,
);

const MistralBedrockResponseCodec = t.type({
  outputs: t.array(
    t.type({
      text: t.string,
      stop_reason: t.string,
    }),
  ),
});

/**
 * @category Responses
 * @category Mistral Bedrock
 */
export interface MistralBedrockResponse
  extends TypeOf<typeof MistralBedrockResponseCodec> {}

export function isMistralBedrockResponse(
  response: unknown,
): response is MistralBedrockResponse {
  return !isLeft(MistralBedrockResponseCodec.decode(response));
}

export interface MistralBedrockApi
  extends ModelApi<MistralBedrockOptions, MistralBedrockResponse> {}

/**
 *
 * ## Reference
 * [Mistral on AWS Bedrock](https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-mistral.html)
 *
 * ## Providers using this API
 * - {@link createAwsBedrockModelProvider | AWS Bedrock}
 *
 * @category APIs
 * @category Mistral Bedrock
 * @category Provider: AWS Bedrock
 *
 */
export const MistralBedrockApi: MistralBedrockApi = {
  requestTemplate: MistralBedrockTemplate,
  responseGuard: isMistralBedrockResponse,
};
