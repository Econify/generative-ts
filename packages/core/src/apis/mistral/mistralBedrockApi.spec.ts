import { MistralBedrockApi, MistralBedrockOptions } from "./mistralBedrockApi";

function render(context: Omit<MistralBedrockOptions, "modelId">) {
  try {
    const rendered = MistralBedrockApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("MistralBedrockApi:", () => {
  /**
   * FewShotRequestOptions ($prompt, examplePairs, system):
   */

  test("$prompt", () => {
    const rendered = render({
      $prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, system", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, system", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
      max_tokens: 1000,
      stop: ["mock-stop-1"],
      temperature: 0.7,
      top_p: 0.9,
      top_k: 50,
    });

    expect(rendered).toMatchSnapshot();
  });
});
