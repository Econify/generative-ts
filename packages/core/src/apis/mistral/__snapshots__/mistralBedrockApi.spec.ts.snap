// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MistralBedrockApi: all options 1`] = `
{
  "max_tokens": 1000,
  "prompt": "<s>[INST] <<SYS>>
mock-system-text
<</SYS>>

mock-user-msg-1 [/INST] mock-assistant-msg-1 </s><s>[INST] mock-prompt [/INST]",
  "stop": [
    "mock-stop-1",
  ],
  "temperature": 0.7,
  "top_k": 50,
  "top_p": 0.9,
}
`;

exports[`MistralBedrockApi: prompt 1`] = `
{
  "prompt": "<s>[INST] mock-prompt [/INST]",
}
`;

exports[`MistralBedrockApi: prompt, examplePairs 1`] = `
{
  "prompt": "<s>[INST] mock-user-msg-1 [/INST] mock-assistant-msg-1 </s><s>[INST] mock-user-msg-2 [/INST] mock-assistant-msg-2 </s><s>[INST] mock-prompt [/INST]",
}
`;

exports[`MistralBedrockApi: prompt, examplePairs, system 1`] = `
{
  "prompt": "<s>[INST] <<SYS>>
mock-system-text
<</SYS>>

mock-user-msg-1 [/INST] mock-assistant-msg-1 </s><s>[INST] mock-user-msg-2 [/INST] mock-assistant-msg-2 </s><s>[INST] mock-prompt [/INST]",
}
`;

exports[`MistralBedrockApi: prompt, system 1`] = `
{
  "prompt": "<s>[INST] <<SYS>>
mock-system-text
<</SYS>>

mock-prompt [/INST]",
}
`;
